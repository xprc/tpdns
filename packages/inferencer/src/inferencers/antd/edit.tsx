import * as RefineAntd from "@pankod/refine-antd";
import dayjs from "dayjs";

import { createInferencer } from "@/create-inferencer";
import {
    jsx,
    componentName,
    prettyString,
    accessor,
    printImports,
    isIDKey,
    noOp,
    getVariableName,
} from "@/utilities";

import { ErrorComponent } from "./error";
import { LoadingComponent } from "./loading";
import { CodeViewerComponent } from "./code-viewer";

import { InferencerResultComponent, InferField, ImportElement } from "@/types";
import { shouldDotAccess } from "@/utilities/accessor";

/**
 * @experimental This is an experimental component
 */
export const EditInferencer: InferencerResultComponent = createInferencer({
    type: "edit",
    additionalScope: [
        ["@pankod/refine-antd", "RefineAntd", RefineAntd],
        ["dayjs", "dayjs", dayjs, true],
    ],
    codeViewerComponent: CodeViewerComponent,
    loadingComponent: LoadingComponent,
    errorComponent: ErrorComponent,
    renderer: ({ resource, fields, isCustomPage, id }) => {
        const COMPONENT_NAME = componentName(
            resource.label ?? resource.name,
            "edit",
        );
        const recordName = getVariableName(
            resource.label ?? resource.name,
            "Data",
        );
        const imports: Array<ImportElement> = [
            ["React", "react", true],
            ["IResourceComponentsProps", "@pankod/refine-core"],
            ["Edit", "@pankod/refine-antd"],
            ["Form", "@pankod/refine-antd"],
            ["useForm", "@pankod/refine-antd"],
            ["Input", "@pankod/refine-antd"],
        ];

        const relationFields: (InferField | null)[] = fields.filter(
            (field) => field?.relation && !field?.fieldable && field?.resource,
        );

        const relationHooksCode = relationFields
            .filter(Boolean)
            .map((field) => {
                if (field?.relation && !field.fieldable && field.resource) {
                    imports.push(["useSelect", "@pankod/refine-antd"]);

                    let val = accessor(
                        recordName,
                        field.key,
                        field.accessor,
                        false,
                    );

                    if (field.multiple && field.accessor) {
                        val = `${accessor(
                            recordName,
                            field.key,
                        )}?.map((item: any) => ${accessor(
                            "item",
                            undefined,
                            field.accessor,
                        )})`;
                    }

                    return `
                    const { selectProps: ${getVariableName(
                        field.key,
                        "SelectProps",
                    )} } =
                    useSelect({
                        resource: "${field.resource.name}",
                        defaultValue: ${val},
                        ${
                            field.relationInfer
                                ? field.relationInfer.accessor
                                    ? typeof field.relationInfer.accessor ===
                                      "string"
                                        ? field.relationInfer.accessor !==
                                          "title"
                                            ? `optionLabel: "${field.relationInfer.accessor}",`
                                            : ""
                                        : `optionLabel: "${field.relationInfer.accessor[0]}",`
                                    : ""
                                : ""
                        }
                    });
                `;
                }
                return undefined;
            })
            .filter(Boolean);

        const renderRelationFields = (field: InferField) => {
            if (field.relation && field.resource) {
                imports.push(["Select", "@pankod/refine-antd"]);
                const variableName = getVariableName(field.key, "SelectProps");

                const name = field.accessor
                    ? field.multiple
                        ? `"${field.key}"`
                        : `["${field.key}", "${field.accessor}"]`
                    : `"${field.key}"`;

                let valueProps = "";
                let valueEvent = "";

                if (field.accessor && field.multiple) {
                    const canDot = shouldDotAccess(`${field.accessor}`);
                    valueEvent = `getValueFromEvent={(selected: string[]) => {
                        return selected?.map((item) => ({ ${
                            canDot ? field.accessor : `["${field.accessor}"]`
                        }: item }));
                    }}`;
                    valueProps = `getValueProps={(value: any[]) => {
                        return {
                            value: value?.map((item) => ${accessor(
                                "item",
                                undefined,
                                field.accessor,
                            )}),
                        };
                    }}`;
                }

                return jsx`
                    <Form.Item
                        label="${prettyString(field.key)}"
                        name={${name}}
                        rules={[
                            {
                                required: true,
                            },
                        ]}
                        ${valueProps}
                        ${valueEvent}
                    >
                        <Select ${
                            field.multiple ? 'mode="multiple"' : ""
                        } {...${variableName}} />
                    </Form.Item>             
                    `;
            }
            return undefined;
        };

        const basicInputFields = (field: InferField) => {
            if (
                field.type === "text" ||
                field.type === "url" ||
                field.type === "email" ||
                field.type === "number"
            ) {
                if (field.multiple) {
                    const val = accessor(field.key, "index", field.accessor)
                        .split("?.")
                        .map((el) => `"${el}"`)
                        .join(", ")
                        .replace(/"index"/, "index");

                    return `
                        <>
                            {(${accessor(
                                recordName,
                                field.key,
                            )} as any[])?.map((item, index) => (
                                <Form.Item
                                    key={index}
                                    label={\`${prettyString(
                                        field.key,
                                    )} \${index+1}\`}
                                    name={[${val}]}
                                >
                                    <Input
                                        type="${field.type}"
                                    ${
                                        isIDKey(field.key)
                                            ? "readOnly disabled"
                                            : ""
                                    } />
                                </Form.Item>
                            ))}
                        </>
                    `;
                }
                return jsx`
                    <Form.Item
                        label="${prettyString(field.key)}"
                        name={["${field.key}"${
                    field.accessor ? ', "' + field.accessor + '"' : ""
                }]}
                        rules={[
                            {
                                required: true,
                            },
                        ]}
                    >
                        <Input ${
                            isIDKey(field.key) ? "readOnly disabled" : ""
                        } />
                    </Form.Item>
                `;
            }
            return undefined;
        };

        const imageFields = (field: InferField) => {
            if (field.type === "image") {
                imports.push(
                    ["Upload", "@pankod/refine-antd"],
                    ["getValueFromEvent", "@pankod/refine-antd"],
                );
                let valueProps = 'valuePropName="fileList"';

                if (field.multiple && !field.accessor) {
                    valueProps =
                        "getValueProps={(value) => ({ fileList: value?.map((item: any) => ({ url: item, name: item, uid: item }))})}";
                }

                if (!field.multiple) {
                    if (field.accessor) {
                        valueProps =
                            "getValueProps={(value) => ({ fileList: value ? [value] : [] })}";
                    } else {
                        valueProps =
                            "getValueProps={(value) => ({ fileList: [{ url: value, name: value, uid: value }]})}";
                    }
                }

                return jsx`
                    <Form.Item label="${prettyString(field.key)}">
                        <Form.Item
                            name="${field.key}"
                            ${valueProps}
                            getValueFromEvent={getValueFromEvent}
                            noStyle
                            rules={[
                                {
                                    required: true,
                                },
                            ]}
                        >
                            <Upload.Dragger
                                listType="picture"
                                ${field.multiple ? "multiple" : ""}
                                beforeUpload={() => false}
                            >
                                <p className="ant-upload-text">
                                    Drag & drop a file in this area
                                </p>
                            </Upload.Dragger>
                        </Form.Item>
                    </Form.Item>
                    `;
            }
            return undefined;
        };

        const booleanFields = (field: InferField) => {
            if (field.type === "boolean") {
                imports.push(["Checkbox", "@pankod/refine-antd"]);

                if (field.multiple) {
                    const val = accessor(field.key, "index", field.accessor)
                        .split("?.")
                        .map((el) => `"${el}"`)
                        .join(", ")
                        .replace(/"index"/, "index");

                    return `
                        <>
                            {(${accessor(
                                recordName,
                                field.key,
                            )} as any[])?.map((item, index) => (
                                <Form.Item
                                    key={index}
                                    valuePropName="checked"
                                    label={\`${prettyString(
                                        field.key,
                                    )} \${index+1}\`}
                                    name={[${val}]}
                                >
                                    <Checkbox>${prettyString(
                                        field.key,
                                    )}</Checkbox>
                                </Form.Item>
                            ))}
                        </>
                    `;
                }
                return jsx`
                    <Form.Item
                        label="${prettyString(field.key)}"
                        valuePropName="checked"
                        name={["${field.key}"${
                    field.accessor ? ', "' + field.accessor + '"' : ""
                }]}
                        rules={[
                            {
                                required: true,
                            },
                        ]}
                    >
                        <Checkbox>${prettyString(field.key)}</Checkbox>
                    </Form.Item>
                `;
            }
            return undefined;
        };

        const dateFields = (field: InferField) => {
            if (field.type === "date") {
                imports.push(
                    ["DatePicker", "@pankod/refine-antd"],
                    ["dayjs", "dayjs", true],
                );

                if (field.multiple) {
                    const val = accessor(field.key, "index", field.accessor)
                        .split("?.")
                        .map((el) => `"${el}"`)
                        .join(", ")
                        .replace(/"index"/, "index");

                    return jsx`
                        <>
                            {(${accessor(
                                recordName,
                                field.key,
                            )} as any[])?.map((item, index) => (
                                <Form.Item
                                    key={index}
                                    label={\`${prettyString(
                                        field.key,
                                    )} \${index+1}\`}
                                    name={[${val}]}
                                    getValueProps={(value) => ({ value: value ? dayjs(value) : undefined })}
                                >
                                    <DatePicker />
                                </Form.Item>
                            ))}
                        </>
                    `;
                }
                return jsx`
                    <Form.Item
                        label="${prettyString(field.key)}"
                        name={["${field.key}"${
                    field.accessor ? ', "' + field.accessor + '"' : ""
                }]}
                        rules={[
                            {
                                required: true,
                            },
                        ]}
                        getValueProps={(value) => ({ value: value ? dayjs(value) : undefined })}
                    >
                        <DatePicker />
                    </Form.Item>
                `;
            }
            return undefined;
        };

        const richtextFields = (field: InferField) => {
            if (field.type === "richtext") {
                return jsx`
                <Form.Item
                    label="${prettyString(field.key)}"
                    name="${field.key}"
                    rules={[
                        {
                            required: true,
                        },
                    ]}
                >
                    <Input.TextArea rows={5} />
                </Form.Item>
                `;
            }

            return undefined;
        };

        const renderedFields: Array<string | undefined> = fields.map(
            (field) => {
                switch (field?.type) {
                    case "text":
                    case "number":
                    case "email":
                    case "url":
                        return basicInputFields(field);
                    case "richtext":
                        return richtextFields(field);
                    case "image":
                        return imageFields(field);
                    case "date":
                        return dateFields(field);
                    case "boolean":
                        return booleanFields(field);
                    case "relation":
                        return renderRelationFields(field);
                    default:
                        return undefined;
                }
            },
        );

        noOp(imports);

        return jsx`
        ${printImports(imports)}
        
        export const ${COMPONENT_NAME}: React.FC<IResourceComponentsProps> = () => {
            const { formProps, saveButtonProps, queryResult } = useForm(${
                isCustomPage
                    ? `{
                          resource: "${resource.name}",
                          id: ${id},
                          action: "edit",
                      }`
                    : ""
            });
        
            const ${recordName} = queryResult?.data?.data;
        
            ${relationHooksCode}

            return (
                <Edit saveButtonProps={saveButtonProps}>
                    <Form {...formProps} layout="vertical">
                        ${renderedFields.join("")}
                    </Form>
                </Edit>
            );
        };
        `;
    },
});
