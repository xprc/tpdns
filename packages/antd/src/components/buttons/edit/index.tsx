import React from "react";
import { Button } from "antd";
import { EditOutlined } from "@ant-design/icons";
import {
    useCan,
    useNavigation,
    useTranslate,
    useResource,
    useRouterContext,
} from "@pankod/refine-core";
import { RefineButtonTestIds } from "@pankod/refine-ui-types";

import { EditButtonProps } from "../types";

/**
 * `<EditButton>` uses Ant Design's {@link https://ant.design/components/button/ `<Button>`} component.
 * It uses the {@link https://refine.dev/docs/core/hooks/navigation/useNavigation#edit `edit`} method from {@link https://refine.dev/docs/core/hooks/navigation/useNavigation `useNavigation`} under the hood.
 * It can be useful when redirecting the app to the edit page with the record id route of resource}.
 *
 * @see {@link https://refine.dev/docs/ui-frameworks/antd/components/buttons/edit-button} for more details.
 */
export const EditButton: React.FC<EditButtonProps> = ({
    resourceName: propResourceName,
    resourceNameOrRouteName: propResourceNameOrRouteName,
    recordItemId,
    hideText = false,
    accessControl,
    ignoreAccessControlProvider = false,
    children,
    onClick,
    ...rest
}) => {
    const accessControlEnabled =
        accessControl?.enabled ?? !ignoreAccessControlProvider;
    const hideIfUnauthorized = accessControl?.hideIfUnauthorized ?? false;
    const translate = useTranslate();

    const { editUrl: generateEditUrl } = useNavigation();
    const { Link } = useRouterContext();

    const { resourceName, id, resource } = useResource({
        resourceName: propResourceName,
        resourceNameOrRouteName: propResourceNameOrRouteName,
        recordItemId,
    });

    const { data } = useCan({
        resource: resourceName,
        action: "edit",
        params: { id, resource },
        queryOptions: {
            enabled: accessControlEnabled,
        },
    });

    const createButtonDisabledTitle = () => {
        if (data?.can) return "";
        else if (data?.reason) return data.reason;
        else
            return translate(
                "buttons.notAccessTitle",
                "You don't have permission to access",
            );
    };

    const editUrl = generateEditUrl(propResourceName ?? resource.route!, id!);

    if (accessControlEnabled && hideIfUnauthorized && !data?.can) {
        return null;
    }

    return (
        <Link
            to={editUrl}
            replace={false}
            onClick={(e: React.MouseEvent<HTMLElement, MouseEvent>) => {
                if (data?.can === false) {
                    e.preventDefault();
                    return;
                }
                if (onClick) {
                    e.preventDefault();
                    onClick(e);
                }
            }}
        >
            <Button
                icon={<EditOutlined />}
                disabled={data?.can === false}
                title={createButtonDisabledTitle()}
                data-testid={RefineButtonTestIds.EditButton}
                {...rest}
            >
                {!hideText && (children ?? translate("buttons.edit", "Edit"))}
            </Button>
        </Link>
    );
};
