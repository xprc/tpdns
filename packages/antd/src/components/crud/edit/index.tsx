import React from "react";

import { Card, Space, PageHeader, Spin } from "antd";
import {
    useResourceWithRoute,
    useMutationMode,
    useNavigation,
    useTranslate,
    useRouterContext,
    userFriendlyResourceName,
    ResourceRouterParams,
    useRefineContext,
} from "@pankod/refine-core";

import {
    DeleteButton,
    RefreshButton,
    ListButton,
    SaveButton,
    Breadcrumb,
} from "@components";
import { EditProps } from "../types";

/**
 * `<Edit>` provides us a layout for displaying the page.
 * It does not contain any logic but adds extra functionalities like a refresh button.
 *
 * @see {@link https://refine.dev/docs/ui-frameworks/antd/components/basic-views/edit} for more details.
 */
export const Edit: React.FC<EditProps> = ({
    title,
    actionButtons,
    saveButtonProps,
    mutationMode: mutationModeProp,
    recordItemId,
    children,
    deleteButtonProps,
    pageHeaderProps,
    canDelete,
    resource: resourceFromProps,
    isLoading = false,
    dataProviderName,
    breadcrumb: breadcrumbFromProps,
    wrapperProps,
    headerProps,
    contentProps,
    headerButtonProps,
    headerButtons,
    footerButtonProps,
    footerButtons,
    goBack: goBackFromProps,
}) => {
    const translate = useTranslate();
    const { goBack, list } = useNavigation();
    const resourceWithRoute = useResourceWithRoute();

    const { mutationMode: mutationModeContext } = useMutationMode();

    const mutationMode = mutationModeProp ?? mutationModeContext;

    const { useParams } = useRouterContext();

    const {
        resource: routeResourceName,
        action: routeFromAction,
        id: idFromRoute,
    } = useParams<ResourceRouterParams>();

    const resource = resourceWithRoute(resourceFromProps ?? routeResourceName);
    const isDeleteButtonVisible =
        canDelete ?? (resource.canDelete || deleteButtonProps);

    const { options } = useRefineContext();
    const breadcrumb =
        typeof breadcrumbFromProps === "undefined"
            ? options?.breadcrumb
            : breadcrumbFromProps;

    const id = recordItemId ?? idFromRoute;

    const defaultHeaderButtons = (
        <>
            {!recordItemId && (
                <ListButton
                    {...(isLoading ? { disabled: true } : {})}
                    resourceNameOrRouteName={resource.route}
                />
            )}
            <RefreshButton
                {...(isLoading ? { disabled: true } : {})}
                resourceNameOrRouteName={resource.route}
                recordItemId={id}
                dataProviderName={dataProviderName}
            />
        </>
    );

    const defaultFooterButtons = (
        <>
            {isDeleteButtonVisible && (
                <DeleteButton
                    {...(isLoading ? { disabled: true } : {})}
                    mutationMode={mutationMode}
                    onSuccess={() => {
                        list(resource.route ?? resource.name);
                    }}
                    dataProviderName={dataProviderName}
                    {...deleteButtonProps}
                />
            )}
            <SaveButton
                {...(isLoading ? { disabled: true } : {})}
                {...saveButtonProps}
            />
        </>
    );

    return (
        <div {...(wrapperProps ?? {})}>
            <PageHeader
                ghost={false}
                backIcon={goBackFromProps}
                onBack={routeFromAction ? goBack : undefined}
                title={
                    title ??
                    translate(
                        `${resource.name}.titles.edit`,
                        `Edit ${userFriendlyResourceName(
                            resource.label ?? resource.name,
                            "singular",
                        )}`,
                    )
                }
                extra={
                    <Space wrap {...(headerButtonProps ?? {})}>
                        {headerButtons
                            ? typeof headerButtons === "function"
                                ? headerButtons({
                                      defaultButtons: defaultHeaderButtons,
                                  })
                                : headerButtons
                            : defaultHeaderButtons}
                    </Space>
                }
                breadcrumb={
                    typeof breadcrumb !== "undefined" ? (
                        <>{breadcrumb}</> ?? undefined
                    ) : (
                        <Breadcrumb />
                    )
                }
                {...(pageHeaderProps ?? {})}
                {...(headerProps ?? {})}
            >
                <Spin spinning={isLoading}>
                    <Card
                        bordered={false}
                        actions={[
                            <Space
                                key="footer-buttons"
                                wrap
                                style={{
                                    float: "right",
                                    marginRight: 24,
                                }}
                                {...(footerButtonProps ?? {})}
                            >
                                {footerButtons
                                    ? typeof footerButtons === "function"
                                        ? footerButtons({
                                              defaultButtons:
                                                  defaultFooterButtons,
                                          })
                                        : footerButtons
                                    : actionButtons
                                    ? actionButtons
                                    : defaultFooterButtons}
                            </Space>,
                        ]}
                        {...(contentProps ?? {})}
                    >
                        {children}
                    </Card>
                </Spin>
            </PageHeader>
        </div>
    );
};
