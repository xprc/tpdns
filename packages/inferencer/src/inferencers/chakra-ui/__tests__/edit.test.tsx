import React from "react";
import { Route, Routes } from "react-router-dom";

import { render, act, TestWrapper } from "@test";
import { EditInferencer } from "../edit";

describe("ChakraEditInferencer", () => {
    it("should match the snapshot", async () => {
        const Wrapper = TestWrapper({
            routerInitialEntries: ["/posts/edit/11"],
            resources: [
                {
                    name: "posts",
                    list: () => <div>list</div>,
                    edit: EditInferencer,
                },
                {
                    name: "categories",
                },
                {
                    name: "tags",
                },
                { name: "users" },
            ],
        });

        const rendering = render(
            <Wrapper>
                <Routes>
                    <Route
                        path="/:resource/edit/:id"
                        element={<EditInferencer resource="posts" />}
                    />
                </Routes>
            </Wrapper>,
        );

        await act(async () => {
            await new Promise((resolve) => setTimeout(resolve, 2000));
        });
        await act(async () => {
            await new Promise((resolve) => setTimeout(resolve, 2000));
        });
        await act(async () => {
            await new Promise((resolve) => setTimeout(resolve, 2000));
        });

        const node = rendering.asFragment();

        expect(node).toMatchSnapshot();
    });
});
