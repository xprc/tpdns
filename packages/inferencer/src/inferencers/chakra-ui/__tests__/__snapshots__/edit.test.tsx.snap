// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChakraEditInferencer should match the snapshot 1`] = `
<DocumentFragment>
  <div>
    <div
      class="css-1mdbftt"
    >
      <div
        class="css-15izviv"
      >
        <div
          class="css-gqssyb"
        >
          <div
            class="chakra-stack css-15lgeks"
          >
            <button
              aria-label="back"
              class="chakra-button css-p7mkev"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="icon icon-tabler icon-tabler-arrow-left"
                fill="none"
                focusable="false"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M0 0h24v24H0z"
                  fill="none"
                  stroke="none"
                />
                <line
                  x1="5"
                  x2="19"
                  y1="12"
                  y2="12"
                />
                <line
                  x1="5"
                  x2="11"
                  y1="12"
                  y2="18"
                />
                <line
                  x1="5"
                  x2="11"
                  y1="12"
                  y2="6"
                />
              </svg>
            </button>
            <h3
              class="chakra-heading css-0"
            >
              Edit Post
            </h3>
          </div>
        </div>
        <div
          class="css-i08qke"
        >
          <a
            href="/posts"
          >
            <button
              class="chakra-button css-4xx2wk"
              title=""
              type="button"
            >
              <span
                class="chakra-button__icon css-1wh2kri"
              >
                <svg
                  aria-hidden="true"
                  class="icon icon-tabler icon-tabler-list"
                  fill="none"
                  focusable="false"
                  height="20"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  width="20"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M0 0h24v24H0z"
                    fill="none"
                    stroke="none"
                  />
                  <line
                    x1="9"
                    x2="20"
                    y1="6"
                    y2="6"
                  />
                  <line
                    x1="9"
                    x2="20"
                    y1="12"
                    y2="12"
                  />
                  <line
                    x1="9"
                    x2="20"
                    y1="18"
                    y2="18"
                  />
                  <line
                    x1="5"
                    x2="5"
                    y1="6"
                    y2="6.01"
                  />
                  <line
                    x1="5"
                    x2="5"
                    y1="12"
                    y2="12.01"
                  />
                  <line
                    x1="5"
                    x2="5"
                    y1="18"
                    y2="18.01"
                  />
                </svg>
              </span>
              Posts
            </button>
          </a>
          <button
            class="chakra-button css-4xx2wk"
            type="button"
          >
            <span
              class="chakra-button__icon css-1wh2kri"
            >
              <svg
                aria-hidden="true"
                class="icon icon-tabler icon-tabler-refresh"
                fill="none"
                focusable="false"
                height="20"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="20"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M0 0h24v24H0z"
                  fill="none"
                  stroke="none"
                />
                <path
                  d="M20 11a8.1 8.1 0 0 0 -15.5 -2m-.5 -4v4h4"
                />
                <path
                  d="M4 13a8.1 8.1 0 0 0 15.5 2m.5 4v-4h-4"
                />
              </svg>
            </span>
            Refresh
          </button>
        </div>
      </div>
      <div
        class="css-0"
      >
        <div
          class="chakra-form-control css-h82jag"
          role="group"
        >
          <label
            class="chakra-form__label css-uvho5g"
            for="field-:r5:"
            id="field-:r5:-label"
          >
            Id
          </label>
          <input
            class="chakra-input css-0"
            disabled=""
            id="field-:r5:"
            name="id"
            type="number"
          />
        </div>
        <div
          class="chakra-form-control css-h82jag"
          role="group"
        >
          <label
            class="chakra-form__label css-uvho5g"
            for="field-:r6:"
            id="field-:r6:-label"
          >
            Title
          </label>
          <input
            class="chakra-input css-0"
            id="field-:r6:"
            name="title"
            type="text"
          />
        </div>
        <div
          class="chakra-form-control css-h82jag"
          role="group"
        >
          <label
            class="chakra-form__label css-uvho5g"
            for="field-:r7:"
            id="field-:r7:-label"
          >
            Slug
          </label>
          <input
            class="chakra-input css-0"
            id="field-:r7:"
            name="slug"
            type="text"
          />
        </div>
        <div
          class="chakra-form-control css-h82jag"
          role="group"
        >
          <label
            class="chakra-form__label css-uvho5g"
            for="field-:r8:"
            id="field-:r8:-label"
          >
            Content
          </label>
          <input
            class="chakra-input css-0"
            id="field-:r8:"
            name="content"
          />
        </div>
        <div
          class="chakra-form-control css-h82jag"
          role="group"
        >
          <label
            class="chakra-form__label css-uvho5g"
            for="field-:r9:"
            id="field-:r9:-label"
          >
            Hit
          </label>
          <input
            class="chakra-input css-0"
            id="field-:r9:"
            name="hit"
            type="number"
          />
        </div>
        <div
          class="chakra-form-control css-h82jag"
          role="group"
        >
          <label
            class="chakra-form__label css-uvho5g"
            for="field-:ra:"
            id="field-:ra:-label"
          >
            Category
          </label>
          <div
            class="chakra-select__wrapper css-42b2qy"
          >
            <select
              class="chakra-select css-1ik61og"
              id="field-:ra:"
              name="category.id"
            >
              <option
                value=""
              >
                Select category
              </option>
              <option
                value="16"
              >
                Numquam Saepe Illo
              </option>
              <option
                value="7"
              >
                Sit Natus Voluptatibus
              </option>
            </select>
            <div
              class="chakra-select__icon-wrapper css-162mkon"
            >
              <svg
                aria-hidden="true"
                class="chakra-select__icon"
                focusable="false"
                role="presentation"
                style="width: 1em; height: 1em; color: currentColor;"
                viewBox="0 0 24 24"
              >
                <path
                  d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                  fill="currentColor"
                />
              </svg>
            </div>
          </div>
        </div>
        <div
          class="chakra-form-control css-h82jag"
          role="group"
        >
          <label
            class="chakra-form__label css-uvho5g"
            for="field-:rb:"
            id="field-:rb:-label"
          >
            User
          </label>
          <div
            class="chakra-select__wrapper css-42b2qy"
          >
            <select
              class="chakra-select css-1ik61og"
              id="field-:rb:"
              name="user.id"
            >
              <option
                value=""
              >
                Select user
              </option>
              <option
                value="21"
              >
                Eriberto
              </option>
              <option
                value="49"
              >
                Tressie
              </option>
            </select>
            <div
              class="chakra-select__icon-wrapper css-162mkon"
            >
              <svg
                aria-hidden="true"
                class="chakra-select__icon"
                focusable="false"
                role="presentation"
                style="width: 1em; height: 1em; color: currentColor;"
                viewBox="0 0 24 24"
              >
                <path
                  d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                  fill="currentColor"
                />
              </svg>
            </div>
          </div>
        </div>
        <div
          class="chakra-form-control css-h82jag"
          role="group"
        >
          <label
            class="chakra-form__label css-uvho5g"
            for="field-:rc:"
            id="field-:rc:-label"
          >
            Status
          </label>
          <input
            class="chakra-input css-0"
            id="field-:rc:"
            name="status"
            type="text"
          />
        </div>
        <div
          class="chakra-form-control css-h82jag"
          role="group"
        >
          <label
            class="chakra-form__label css-uvho5g"
            for="field-:rd:"
            id="field-:rd:-label"
          >
            Status Color
          </label>
          <input
            class="chakra-input css-0"
            id="field-:rd:"
            name="status_color"
            type="text"
          />
        </div>
        <div
          class="chakra-form-control css-h82jag"
          role="group"
        >
          <label
            class="chakra-form__label css-uvho5g"
            for="field-:re:"
            id="field-:re:-label"
          >
            Created At
          </label>
          <input
            class="chakra-input css-0"
            id="field-:re:"
            name="createdAt"
          />
        </div>
        <div
          class="chakra-form-control css-h82jag"
          role="group"
        >
          <label
            class="chakra-form__label css-uvho5g"
            for="field-:rf:"
            id="field-:rf:-label"
          >
            Published At
          </label>
          <input
            class="chakra-input css-0"
            id="field-:rf:"
            name="publishedAt"
          />
        </div>
        <div
          class="chakra-form-control css-h82jag"
          role="group"
        >
          <label
            class="chakra-form__label css-uvho5g"
            for="field-:rg:"
            id="field-:rg:-label"
          >
            Tags
          </label>
          <div
            class="chakra-select__wrapper css-42b2qy"
          >
            <select
              class="chakra-select css-1ik61og"
              id="field-:rg:"
              name="tags"
            >
              <option
                value=""
              >
                Select tag
              </option>
              <option
                value="5"
              >
                Error Eos Et
              </option>
              <option
                value="9"
              >
                Nobis Et Ut
              </option>
            </select>
            <div
              class="chakra-select__icon-wrapper css-162mkon"
            >
              <svg
                aria-hidden="true"
                class="chakra-select__icon"
                focusable="false"
                role="presentation"
                style="width: 1em; height: 1em; color: currentColor;"
                viewBox="0 0 24 24"
              >
                <path
                  d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                  fill="currentColor"
                />
              </svg>
            </div>
          </div>
        </div>
        <div
          class="chakra-form-control css-h82jag"
          role="group"
        >
          <label
            class="chakra-form__label css-uvho5g"
            for="field-:rh:"
            id="field-:rh:-label"
          >
            Language
          </label>
          <input
            class="chakra-input css-0"
            id="field-:rh:"
            name="language"
            type="number"
          />
        </div>
      </div>
      <div
        class="css-e6thd7"
      >
        <button
          class="chakra-button css-4xx2wk"
          type="button"
        >
          <span
            class="chakra-button__icon css-1wh2kri"
          >
            <svg
              aria-hidden="true"
              class="icon icon-tabler icon-tabler-device-floppy"
              fill="none"
              focusable="false"
              height="20"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="20"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M0 0h24v24H0z"
                fill="none"
                stroke="none"
              />
              <path
                d="M6 4h10l4 4v10a2 2 0 0 1 -2 2h-12a2 2 0 0 1 -2 -2v-12a2 2 0 0 1 2 -2"
              />
              <circle
                cx="12"
                cy="14"
                r="2"
              />
              <polyline
                points="14 4 14 8 8 8 8 4"
              />
            </svg>
          </span>
          Save
        </button>
      </div>
    </div>
  </div>
  <div
    class="css-1x21b2v"
  >
    <div
      class="chakra-stack css-1qwdsol"
    >
      <button
        class="chakra-button css-4xx2wk"
        type="button"
      >
        <span
          class="chakra-button__icon css-1wh2kri"
        >
          <svg
            aria-hidden="true"
            class="icon icon-tabler icon-tabler-code"
            fill="none"
            focusable="false"
            height="18"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="18"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
              stroke="none"
            />
            <polyline
              points="7 8 3 12 7 16"
            />
            <polyline
              points="17 8 21 12 17 16"
            />
            <line
              x1="14"
              x2="10"
              y1="4"
              y2="20"
            />
          </svg>
        </span>
        Show Code
      </button>
      <a
        class="chakra-button css-4xx2wk"
        href="https://github.com/refinedev/refine/discussions/3046"
        target="_blank"
      >
        <span
          class="chakra-button__icon css-1wh2kri"
        >
          <svg
            aria-hidden="true"
            class="icon icon-tabler icon-tabler-message-circle"
            fill="none"
            focusable="false"
            height="18"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="18"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
              stroke="none"
            />
            <path
              d="M3 20l1.3 -3.9a9 8 0 1 1 3.4 2.9l-4.7 1"
            />
            <line
              x1="12"
              x2="12"
              y1="12"
              y2="12.01"
            />
            <line
              x1="8"
              x2="8"
              y1="12"
              y2="12.01"
            />
            <line
              x1="16"
              x2="16"
              y1="12"
              y2="12.01"
            />
          </svg>
        </span>
        Give Feedback
      </a>
    </div>
  </div>
  <div
    style="height: 50px;"
  />
</DocumentFragment>
`;
